### ISSUE TITLE
Kubernetes not working on EC2 because access to kube-prox is blocked

### ISSUE THREAD

**Issue by jondubois** [rancher/rancher#4959]
**Rancher Version:**

latest

**Docker Version:**

1.10.3

**OS and where are the hosts located? (cloud, bare metal, etc):**

Ubuntu, Amazon EC2

**Setup Details: (single node rancher vs. HA rancher, internal DB vs. external DB)**

Two nodes - Agent runs on a different node from rancher/server - Agent instance was launched through the 'add host' UI and choosing EC2 option.

**Environment Type: (Cattle/Kubernetes/Swarm/Mesos)**

Kubernetes

**Steps to Reproduce:**

1. Have rancher/server running from an instance on EC2 in any availability zone and serve up the control panel.
2. From the infrastructure screen on the Rancher control panel, add a host - Choose EC2.
3. Provide the access keys, fill in all the details, select the VPC option, select `rancher-machine` security group, launch in same availability zone as rancher/server host
4. Once all the forms have been filled out, add the new host
5. Deploy a simple Docker image (e.g. a lightweight/basic hello world app/service) to the cluster using `kubectl`, expose on any port  - In my case the command looked like (image from DockerHub): `kubectl run hello-node --image=myaccount/hello-node:v1 --port=8000`
6. Now I can see my `hello-node` service. Also the container shows up on my control panel as normal (everything looks good so far).
7. Once it's running, expose the service ` kubectl expose deployment hello-node --type="LoadBalancer"`
8. Now I run `kubectl get services` - I can see my `hello-node` service there - Which is good but the EXTERNAL-IP is empty.
9. EXTERNAL-IP is still empty even after 20 minutes...
10. When I try to interact with my app from my browser using the EC2 host IP address (on port 8000), it just gets stuck loading
11. When I change my security group settings to allow all traffic, it works and I can access my service as expected.

**Results:** 

It turns out that traffic to port `30889` was blocked by my EC2 security group (I'm using the default `rancher-machine` security group - I'm not sure if 30889 was randomly allocated?). `kube-prox` is the process which was attached (on the receiving end) of that port. What's strange is that the remoteAddress is the Elastic IP address of that same host - Basically, whatever process (which is running on the same host as `kube-prox`) is trying to talk to `kube-prox` is not using an internal private IP address; it uses the public elastic address to refer to itself.

I tried modifying the `rancher-machine` security group to allow all traffic from all hosts which have `rancher-machine` as their security group (EC2 lets you specify the security group ID instead of an address) - This would have worked if a private IP address was used, but it doesn't work because the EC2 security group inbound rule only considers private IP addresses (not public elastic addresses) when deciding whether to allow or block connections related to a security group.

One potential (naive) solution I can think of:

1. Change the default inbound rules for the EC2 `rancher-machine` security group so that it automatically allows all traffic from other `rancher-machine` hosts
2. Instead of trying to connect to `kube-prox` using its public elastic IP address, we could use its private IP addresses within the VPC/subnet (that way EC2 would not block it)

Another naive solution would be to expose port `30889` to all addresses (though I don't know the security implications of this).

Maybe there are better solutions? Has anyone else encountered similar issues?

**Comment by aemneina** [comment]
Hi @jondubois,

You can change the IP address of the hosts, to use the internal IP address on the VPC. When adding hosts, you can modify the docker run command to force the IP address you want communication bound to. This is done by including the environmental variable ```CATTLE_AGENT_IP=<INTERNAL_IP_OF_HOST>```.

```
sudo docker run -d -e CATTLE_AGENT_IP=<INTERNAL_IP_OF_HOST> --privileged -v /var/run/docker.sock:/var/run/docker.sock -v /var/lib/rancher:/var/lib/rancher rancher/agent:v1.0.1 http://<TO_THE_RANCH>:8080/v1/scripts/...
```

I still had to allow traffic on the VPC, for internal communication, via security groups.

**Comment by jondubois** [comment]
@aemneina Is it possible to do this when adding multiple hosts through Amazon EC2 (from the dashboard)?

**Comment by jondubois** [comment]
@aemneina When creating the hosts through the EC2 hosts menu, I can see a 'Use only private IP address' checkbox - Do I need to tick this? 

![screenshot from 2016-06-01 03-05-57](https://cloud.githubusercontent.com/assets/1187447/15683447/e1bcf1a6-27a5-11e6-8c9a-79eb510488b7.png)

I did try ticking the 'Use only private IP address' checkbox, but because my rancher/server host was not part of the VPC, my hosts would just hang on my Rancher control panel and never launch.

To try to get around the issue, I added a 'VPC ClassicLink' to link my rancher/server host to my target VPC (apparently, this should give rancher/server access to private IPs which are on that VPC) but when I tried launching hosts again, I got the same result - They would get stuck on my Rancher dashboard.

What are the best settings/configurations to use when launching hosts on EC2 with Kubernetes so that it ALL works straight away inside a VPC without exposing any sensitive ports? I can't find any documentation about how to achieve this.

The only way I was able to get Kubernetes working on EC2 was by exposing the `30889` port publicly (which isn't great).

**Comment by violettine** [comment]
I have exactly the same issue! Is there any non-security-risk solution at hand now??

My settings are identical to @jondubois except I am using Docker version 1.11.0 .

**Comment by jondubois** [comment]
@violettine My solution at the moment is to not use the default LoadBalancer and instead to deploy my own load balancer containers (e.g. HAProxy or nginx) as a separate service (without using the `kubectl expose --type="LoadBalancer"` command). This gives me more control so I can scale up/down the loadbalancer service independently.

**Comment by aemneina** [comment]
@jondubois sorry for the delay in response. Using the private address space sounds like the way you want to go. That is achievable with that check box you cite. Your rancher server will need to be able to ssh directly to that host, so your rancher server will need to be able to route to that private subnet (probably the reason you saw the failed launch).

**Comment by deniseschannon** [comment]
@jondubois Please re-open if you have further questions.

@violettine FYI, Kubernetes only supports Docker 1.10.3
